Feature: frontend paperless mail

shamoon opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

This PR (draft pending beta) brings the settings for paperless mail to the frontend UI. It will go nicely (but doesnt depend on) another PR we're cooking up for frontend permissions. With this I also am proposing we move the 'Admin' link to within settings, especially once permissions are in place, I think accessing the admin will be even less frequent and the link can safely be moved. Screenshots below

Edit: this also contains support for routing to specific settings tabs, e.g. `/settings/mail`

I added new tests for the API endpoints. Also note the mail_accounts endpoint returns the imap password obfuscated, i.e. just *** but can accept a new password.

<img width="1391" alt="Screen Shot 2022-11-18 at 5 10 52 PM" src="https://user-images.githubusercontent.com/4887959/202826953-4e5fccb8-6cb1-404c-b8ba-6d358400cd12.png">
<img width="1157" alt="Screen Shot 2022-11-18 at 5 06 12 PM" src="https://user-images.githubusercontent.com/4887959/202826951-4bca34cd-79d1-4b2b-8fe4-d420e2d71069.png">
<img width="1157" alt="Screen Shot 2022-11-18 at 5 02 17 PM" src="https://user-images.githubusercontent.com/4887959/202826955-e218fed1-9377-42f3-b0c0-75719030a69d.png">

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

shamoon committed:
Basic data retrieval

shamoon committed:
remove unused toastService from edit dialogs and add confirmation

shamoon committed:
Mail account edit dialog

shamoon committed:
Mail rule edit dialog

shamoon committed:
mail account + rule deletion

shamoon committed:
Working mail rule & account edit

shamoon committed:
Update deprecated edit-dialog rxjs

shamoon committed:
Update frontend fixtures & tests for compatibility

shamoon committed:
Mail form tweaks

Include add button

Include add button

shamoon committed:
Update welcome tour, move admin button

shamoon committed:
dynamic loading of settings tab contents

shamoon committed:
Dynamically load mail rules / accounts settings

shamoon committed:
update settings tests to not wait on data which is now on-demand

shamoon committed:
Fix mail account / rule delete

shamoon committed:
one-way imap password setting via API, ObfuscatedPasswordField

shamoon committed:
API mail rule & account tests

and fix use of assign_tags

shamoon committed:
Hide order parameter, fix imap port

shamoon committed:
Update frontend strings

shamoon committed:
Add settings routing

stumpylog commented on src/documents/tests/test_api.py:
```diff
@@ -2929,3 +2931,441 @@ def test_task_name_consume_folder(self):
+        self.assertEqual(response.data["count"], 1)
+        returned_account1 = response.data["results"][0]
+
+        from pprint import pprint
```
Probably don't need the printing now in these cases?

stumpylog reviewed (changes requested):
It looks nice.  Just a couple things.

I get `UnorderedObjectListWarning` for `MailAccount` when loading the mail tab.  Looks like the fix is setting an ordering for the model.

I wasn't able to select a value from the dropdowns in the rule window.  I could drop it down to see the options, but not select after that.

stumpylog commented on src/documents/serialisers.py:
```diff
@@ -688,3 +690,98 @@ def _validate_task_id_list(self, tasks, name="tasks"):
+        return MailAccount.objects.all()
+
+
+class MailRuleSerializer(serializers.ModelSerializer):
```
I'm not sure if this is the right place, but can we validate the inputs some?  Like a MOVE action with an empty parameter, or TAG with an empty parameter, etc

shamoon replied:
Im not sure about this one, we get validation from the model and basic DRF stuff, do you think more is needed? Obviously please feel free to add here as Im not 100% sure what you mean

shamoon replied:
Oh I gotcha now!

shamoon committed:
Apply code suggestions from @stumpylog

shamoon committed:
fix edit dialog getters

shamoon commented:
Whoops, fixed the dropdown thing. As for `UnorderedObjectListWarning` Im not seeing that (not sure why?) I added `order_by(id)` on the `AccountField` `get_queryset` but Im not positive if thats where its needed (since Im not seeing the warning). LMK!

shamoon committed:
frontend mail rule validation

Display non-field validation errors, hide action param field if not needed

shamoon commented:
Ok cool, great suggestions. I added validation but I also made the UI hide the action param field if its not needed. Double validation =)

<img width="1148" alt="Screen Shot 2022-11-28 at 3 55 42 PM" src="https://user-images.githubusercontent.com/4887959/204405507-7c446d48-bb53-4cd1-812e-a61013026568.png">
<img width="1158" alt="Screen Shot 2022-11-28 at 3 55 54 PM" src="https://user-images.githubusercontent.com/4887959/204405509-55b717a5-fbe7-44eb-a0cc-a0099f02c079.png">

shamoon committed:
Re-org where some of the new classes are found

shamoon commented:
sorry for all the psychotic force-pushing, trying to figure out what the heck I did to the commits here 😵

shamoon committed:
prevent loss of unsaved changes to settings on tab nav

shamoon commented:
Ok, sheesh. I've managed to remedy whatever mistake I made to explode the commits and fixed another small issue with saving changes.

I think this should still be good to go, appreciate all your help stumpy, as usual.

Only question (unless you notice anything else) is whether you're still seeing the `UnorderedObjectListWarning`?

stumpylog commented:
I'll grab the latest container and see if I see it again.

Should api.rst be updated to describe the new endpoints?

shamoon commented:
Good call, added the endpoints to the api doc

stumpylog commented on src/documents/views.py:
```diff
@@ -910,3 +914,37 @@ def post(self, request, *args, **kwargs):
+class MailAccountViewSet(ModelViewSet):
+    model = MailAccount
+
+    queryset = MailAccount.objects.all()
```
I think this is where the warning can be fixed,  `order_by("pk")` I think?  It might also be able to put in the `class Meta:` for each model to set to more globally.

```
    class Meta:
        ordering = ("name",)
```

shamoon replied:
Ah ok, thanks, I do see others (though not 1) have explicit ordering params. OK let me know if youre still seeing it with the last commit!

shamoon committed:
Explicit default ordering for rule / account views

stumpylog committed:
Moves where the mail views live and puts the ordering on those

stumpylog reviewed (approved):
Tweaked where those sorting options went, looks like that change did fix it.  I guess the next feature release is the email one

shamoon commented:
Awesome thanks. And whoops looks like in my force-commit debacle I un-moved them thanks.

yes I have another big PR cooked up, we can decide whether to release them together or not later. I’m gonna wait a day or two before committing this just in case we need another bugfix

Thanks as always for the collab!

shamoon commented:
Nvm about waiting, with the .eml PR in there it’s moot :)

morremeyer commented:
Opened #2265 to track adding the **consumption scope** from #848 to the frontend as they were released together.

This should prevent confusion by other users since they might expect that to already be available.
